### DISCUSSION TITLE
My CI is not seeing my environment variables!

### DISCUSSION THREAD

**Discussion by thorwhalen** [i2mint/isee#24] [Q&A] [answered]
My tests are failing because the code doesn't see my environment variables, though I do have them in my "repository secrets".

Concrete example:

This part fails:

```
      - name: Pytest Validation
        uses: i2mint/isee/actions/pytest-validation@master
        with:
          root-dir: ${{ env.PROJECT_NAME }}
          paths-to-ignore: examples,scrap
```

where (currently) the `` contents [can be found here](https://github.com/i2mint/isee/blob/e3f69325562dfdc1aec2a8626d6862975c3a57b3/actions/pytest-validation/action.yml)

**Comment by thorwhalen** [comment, marked as answer]
## tldr;

Add this to the part of the CI that fails:

```
        env:
          KAGGLE_KEY: ${{ secrets.KAGGLE_KEY }}
```

## details

To ensure that your GitHub Actions workflow can access the `KAGGLE_KEY` environment variable that you've stored in your repository secrets, you need to explicitly pass this secret to the workflow. Secrets are not passed to the runner environment by default for security reasons. 

Here's how you can modify your `ci.yml` file to include the `KAGGLE_KEY` environment variable:

```yaml
      - name: Pytest Validation
        uses: i2mint/isee/actions/pytest-validation@master
        with:
          root-dir: ${{ env.PROJECT_NAME }}
          paths-to-ignore: examples,scrap
        env:
          KAGGLE_KEY: ${{ secrets.KAGGLE_KEY }}
```

In this snippet, the `env` section under the `Pytest Validation` step is used to set environment variables for that step. Here, `KAGGLE_KEY` is set to the value of the secret `KAGGLE_KEY` from your repository's secrets.

Remember, for this to work:

1. The secret `KAGGLE_KEY` must be correctly set in your repository's settings under the 'Secrets' section.

2. The GitHub Actions runner needs to have the appropriate permissions to access repository secrets. This is usually the case by default, but if you're using a self-hosted runner or have modified the default permissions, you may need to adjust them.

3. Ensure that your code is correctly configured to use the `KAGGLE_KEY` environment variable.

By making this change, your GitHub Actions workflow should be able to access the `KAGGLE_KEY` environment variable and use it during the `Pytest Validation` step.

**Reply by thorwhalen** [reply]
My ci.yml is actually generated by a template that remains the same for all my projects. In this template, all variations between projects happen via such variables as `${{ env.PROJECT_NAME }}` etc. 

The solution you propose obliges me to edit my copy of the template, which I don't want to do because it makes it harder for me to maintain. 

Is there a way for me to change to include something in the template that will have the effect that anytime the code will look for some environment variable, it will know to look for it in secrets (or some other place that is particular to the repository)?

**Reply by thorwhalen** [reply]
To create a flexible GitHub Actions workflow template that automatically maps environment variables to secrets without requiring per-repository modifications, you can use a dynamic approach to set all secrets as environment variables. However, GitHub Actions does not directly support dynamically mapping all secrets to environment variables due to security reasons. Each secret must be explicitly declared.

That said, there is a workaround using a script within your workflow. You can write a shell script step that reads a predefined list of environment variable names (which are common across your projects) and sets them using the corresponding secrets. Here's an example approach:

1. **Define a list of common environment variable names** in your workflow file. These are the names you expect to use across different projects.

2. **Write a script in your workflow** that loops through these variable names, checks if a secret with the same name exists, and sets an environment variable with its value.

Here's a modified snippet of your `ci.yml` to illustrate this:

```yaml
jobs:
  build:
    runs-on: ubuntu-latest
    env:
      COMMON_ENV_VARS: 'KAGGLE_KEY,ANOTHER_VAR,SOME_OTHER_VAR'
    steps:
      - name: Set Environment Variables from Secrets
        run: |
          for var in ${COMMON_ENV_VARS//,/ }; do
            if [ ! -z "${{ secrets[var] }}" ]; then
              echo "$var=${{ secrets[var] }}" >> $GITHUB_ENV
            fi
          done

      - name: Pytest Validation
        uses: i2mint/isee/actions/pytest-validation@master
        with:
          root-dir: ${{ env.PROJECT_NAME }}
          paths-to-ignore: examples,scrap
        # Now, the environment variables are set globally for this job
```

In this script:

- `COMMON_ENV_VARS` is a comma-separated list of environment variable names.
- The `for` loop iterates through each name, checks if a secret with that name exists, and if so, sets an environment variable with its value.

Remember, this approach requires that the secret names match the environment variable names exactly. Also, be aware that while this method can improve maintenance efficiency, it might pose a security risk by potentially exposing secrets if not handled cautiously.

If this solution does not meet your security standards, or if you have a large number of variable names that vary significantly across projects, you might need to consider a different approach, possibly involving external scripts or more complex GitHub Actions configurations.

**Reply by thorwhalen** [reply]
Still, the solutions above (from chatGPT) are not satisfactory since they all require a list of common environment variables that would cover all projects. As soon as the list is not exhaustive (and it will never be for the general population), one would need to edit the template. 

Instead, I propose that instead of:

```
    env:
      COMMON_ENV_VARS: 'KAGGLE_KEY,ANOTHER_VAR,SOME_OTHER_VAR'
```

we have something like:

```
    env:
     ENV_VARS: ${{ env.ENV_VARS }}
```

so that the user can specify the list of keys in a place that's separate from the template itself, which can now be left un-edited. 

We could also consider using a script/function that would fetch this list automatically/dynamically. Are there any security concerns there?

```python
def get_repo_secret_keys(owner, repo, token=None):
    """Gets list of secret keys for a given repo."""
    import json
    import urllib.request

    url = f"https://api.github.com/repos/{owner}/{repo}/actions/secrets"

    if token is None:
        token = os.environ.get("GITHUB_TOKEN", None)
    headers = {
        "Accept": "application/vnd.github.v3+json"
    }
    if token is not None:
        headers["Authorization"] = f"token {token}"

    req = urllib.request.Request(url, headers=headers)

    with urllib.request.urlopen(req) as response:
        result = json.load(response)
        secret_keys = [secret['name'] for secret in result.get('secrets', [])]
        return secret_keys
```

**Reply by thorwhalen** [reply]
Your approach of using an environment variable to specify the list of secret keys (`ENV_VARS`) is a good idea for maintaining flexibility across different projects without altering the template. This way, you can set `ENV_VARS` in each project's GitHub Actions environment or in the workflow dispatch/event inputs, thereby keeping the template generic.

Regarding the use of a script or function to fetch the list of secret keys dynamically, there are a few security and practical considerations:

1. **Exposure of Sensitive Information:** Automatically fetching and setting all secrets as environment variables might inadvertently expose sensitive information. It's crucial to ensure that only necessary secrets are made available to the workflow.

2. **Access Token Permissions:** The script requires a GitHub token with sufficient permissions (typically admin rights to the repository) to access the secrets. This token must be handled securely.

3. **Token Storage:** If the token is stored as a secret and used within the same workflow to fetch other secrets, it creates a recursive dependency, which can be complex to manage securely.

4. **Execution Context:** The script needs to be executed in a context where it has access to the necessary credentials to authenticate with the GitHub API. This is typically before the main actions that require these environment variables.

5. **Rate Limiting and Error Handling:** When interacting with the GitHub API, consider rate limiting and implement robust error handling to manage potential issues during the API call.

6. **Workflow Complexity:** Adding this level of automation increases the complexity of the workflow. It's important to balance the need for automation with the maintainability and readability of the workflow.

Your modified Python function looks suitable for fetching the list of secret keys. Remember to handle the `GITHUB_TOKEN` securely and ensure that it's only used in contexts where its exposure is controlled. The script should ideally be executed in a step before any other steps that require those environment variables. 

In conclusion, while your approach adds flexibility, it also introduces complexity and potential security considerations. It's crucial to weigh these factors against the benefits of automation in your specific context.

**Reply by thorwhalen** [reply]
See: https://github.com/thorwhalen/haggle/issues/1

**Reply by valentin-feron** [reply]
If the intent is to use `env.SOMETHING` in reusable ci scripts like https://github.com/i2mint/isee/blob/e3f69325562dfdc1aec2a8626d6862975c3a57b3/actions/pytest-validation/action.yml#L1-L42 I don't think we should. We should always use inputs so the scripts do not depend on any environment configuration. Unless there is a case I am missing that would justify such a behavior. In that case, please share it (I don't see any ref to `KAGGLE_KEY` in the latest isee)

**Reply by thorwhalen** [reply]
Let me clarify.

Code sometimes depends on the existence of environment configuration -- for example, secret tokens. 
These usually require manual user input: Either the install instructions tell the user to acquire and specify this information via environment variables, or/and ask the user for this information if/when missing. 

Of course, there's pros and cons to this way of doing things, but we don't really have the choice here, if we're thinking of adapting to a common practice. 

The problem is: Given code whose tests depend on environment variables, how do we communicate these to our Github actions processes. 

It seems you're saying "don't depend on environment variables", which is not a solution to the problem above, just an avoidance of it. The [XY problem](https://en.wikipedia.org/wiki/XY_problem) argument (if that's what's happening here) doesn't apply as I see it. 
Code **will** depend on environmental variables.
